internal/literals/split.go

```diff
@@ -0,0 +1,198 @@
+package literals
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	mathrand "math/rand"
+	ah "mvdan.cc/garble/internal/asthelper"
+)
+
+const maxChunkSize = 4
+const minCaseCount = 3
+
+type split struct{}
+
+// check that the obfuscator interface is implemented
+var _ obfuscator = split{}
+
+func splitIntoRandomChunks(data []byte) [][]byte {
+	if len(data) == 1 {
+		return [][]byte{data}
+	}
+
+	var chunks [][]byte
+	for {
+		if len(data) == 0 {
+			break
+		}
+		chunkSize := 1 + mathrand.Intn(maxChunkSize)
+		if chunkSize > len(data) {
+			chunkSize = len(data)
+		}
+
+		chunks = append(chunks, data[:chunkSize])
+		data = data[chunkSize:]
+	}
+	return chunks
+}
+
+func splitIntoOneByteChunks(data []byte) [][]byte {
+	var chunks [][]byte
+	for _, d := range data {
+		chunks = append(chunks, []byte{d})
+	}
+	return chunks
+}
+
+func shuffleStmts(stmts []ast.Stmt) []ast.Stmt {
+	mathrand.Shuffle(len(stmts), func(i, j int) {
+		stmts[i], stmts[j] = stmts[j], stmts[i]
+	})
+	return stmts
+}
+
+func (x split) obfuscate(data []byte) *ast.BlockStmt {
+	var chunks [][]byte
+	if len(data)/maxChunkSize < minCaseCount { // Short arrays should be divided into single-byte fragments
+		chunks = splitIntoOneByteChunks(data)
+	} else {
+		chunks = splitIntoRandomChunks(data)
+	}
+
+	// Generate indexes for cases chunk count + 1 decrypt case + 1 exit case
+	indexes := mathrand.Perm(len(chunks) + 2)
+
+	decryptKeyInitial := mathrand.Int()
+	decryptKey := decryptKeyInitial
+	// Calculate decrypt key based on indexes and position. Ignore exit index
+	for i, index := range indexes[:len(indexes)-1] {
+		decryptKey ^= index * i
```

Comment:
The idea is different. The key should form only after passing through all switch-case and only in the right order (for this purpose you need a separate counter). This should greatly increase the complexity of automatic analysis.

Reply:
A separate key for each is not very convenient to do. Since the chunks are of different lengths:  https://gist.github.com/pagran/5378b99fa5d5d79cdb7d18b89b3326a7

Reply:
Yes but what if you use  the new key on each element directly, because then equal segments would have different values.

Reply:
https://gist.github.com/pagran/5378b99fa5d5d79cdb7d18b89b3326a7#file-long_string-go-L83 i can add a position in the loop:
```go
for k, y := range data {
	data[y] ^= byte(decryptKey) ^ k
}
```

Reply:
https://gist.github.com/pagran/fc93f088b3be1e797c234b5e1c323d3c

Like this

Reply:
nvm we can keep it this way I didn't take the variable chunk size into account.